Std/Tactic/LetProjs.lean

```diff
@@ -0,0 +1,82 @@
+/-
+Copyright (c) 2023 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import Std.Lean.Meta.Basic
+import Std.Lean.Expr
+import Lean.Elab.Tactic.ElabTerm
+
+/-!
+# `let_projs`
+
+`let_projs` adds let bindings for all projections of local hypotheses, recursively.
+
+It should not be used interactively,
+but may be useful as a preprocessing step for tactics such as `solve_by_elim`.
+-/
+
+open Lean Meta
+
+/--
+Analogue of `Lean.getStructureFields`,
+but return an empty array rather than panicking if `structName` does not refer to a structure.
+-/
+def Lean.getStructureFields! (env : Environment) (structName : Name) : Array Name :=
+  if let some info := getStructureInfo? env structName then
+    info.fieldNames
+  else
+    #[]
+
+namespace Std.Tactic.LetProjs
+
+/--
+Builds all the projections of an expression,
+returning an array of pairs consisting of the projection name
+and the projection applied to the original expression.
+-/
+def allProjs (e : Expr) : MetaM (Array (Name × Expr)) := do
+  let (c, _) := (← inferType e).getAppFnArgs
+  let env ← getEnv
+  unless isStructure env c do return #[]
+  (getStructureFields! env c).filterMapM fun f => return (f, ← mkProjection e f)
```

Comment:
`getStructureFields!" already returns `#[]` if it’s not a structure, so maybe
```suggestion
  (getStructureFields! env c).filterMapM fun f => return (f, ← mkProjection e f)
```
and maybe don’t name it with `!`?

Reply:
I've just inlined it.